src/backend/resourcemanager/resourcemanager.c

```diff
@@ -2177,6 +2213,149 @@ int generateAllocRequestToBroker(void)
+		if (!IS_SEGSTAT_FTSAVAILABLE(segres->Stat) ||
+			(DRMGlobalInstance->ImpType != NONE_HAWQ2 &&
+			 !IS_SEGSTAT_GRMAVAILABLE(segres->Stat)) ||
+			(segres->DecPending.MemoryMB > 0 && segres->DecPending.Core > 0))
+		{
+			index++;
+			continue;
+		}
+
+		int clevel = segres->ContainerSets[0] == NULL ?
+					 0 :
+					 list_length(segres->ContainerSets[0]->Containers) +
+					 segres->IncPending.MemoryMB / ratio;
+
+		ListCell *pcell = NULL;
+		foreach(pcell, *preferred)
+		{
+			PAIR existpair = (PAIR)lfirst(pcell);
+			if ( pair->Value == segres )
+			{
+				reqidx[index] = existpair;
+				totalcount += ((ResourceBundle)(reqidx[index]->Value))->MemoryMB /
+							  ratio;
+				break;
+			}
+		}
+
+
+		int creqsize = reqidx[index] == NULL ?
+					   0 :
+					   ((ResourceBundle)(reqidx[index]->Value))->MemoryMB / ratio;
+
+		llevel = clevel+creqsize < llevel ? clevel+creqsize : llevel;
+		index++;
+	}
+
+	/* Step 2. Adjust request. */
+	int32_t reqcoreleft = *reqcore - totalcount;
+	while( reqcoreleft > 0 )
+	{
+		llevel++;
+		index = 0;
+		foreach(cell, ressegl)
+		{
+			PAIR pair = (PAIR)lfirst(cell);
+			SegResource segres = (SegResource)(pair->Value);
+
+			/*
+			 * Resource manager skips this segment if
+			 * 1) Not FTS available;
+			 * 2) Not GRM available;
+			 * 3) Having resource decrease pending.
+			 */
+			if (!IS_SEGSTAT_FTSAVAILABLE(segres->Stat) ||
+				(DRMGlobalInstance->ImpType != NONE_HAWQ2 &&
+				 !IS_SEGSTAT_GRMAVAILABLE(segres->Stat)) ||
+				(segres->DecPending.MemoryMB > 0 && segres->DecPending.Core > 0))
+			{
+				index++;
+				continue;
+			}
+
+			int clevel = segres->ContainerSets[0] == NULL ?
+						 0 :
+						 list_length(segres->ContainerSets[0]->Containers) +
+						 segres->IncPending.MemoryMB / ratio;
+
+			int aclevel = reqidx[index] == NULL ?
+						  clevel :
+						  clevel + ((ResourceBundle)(reqidx[index]->Value))->MemoryMB / ratio;
+
+			if ( llevel > aclevel )
+			{
+
+				if ( reqidx[index] == NULL )
+				{
+					reqidx[index] = rm_palloc0(PCONTEXT, sizeof(PAIRData));
+					reqidx[index]->Key = segres;
+					ResourceBundle resource = rm_palloc0(PCONTEXT,
+														 sizeof(ResourceBundleData));
+					resetResourceBundleData(resource, 0, 0, ratio);
+					reqidx[index]->Value = resource;
+					*preferred = lappend(*preferred, reqidx[index]);
+				}
+				addResourceBundleData((ResourceBundle)(reqidx[index]->Value),
+									  (llevel-aclevel) * ratio,
+									  llevel-aclevel);
+				reqcoreleft -= llevel-aclevel;
+
+				elog(RMLOG, "Resource manager acquires %lf GRM containers on "
+						    "host %s. Current level(having pending) %d, "
+						    "expect level %d, acquired in current request %d.",
+						    ((ResourceBundle)(reqidx[index]->Value))->Core,
+						    GET_SEGRESOURCE_HOSTNAME(((SegResource)(reqidx[index]->Key))),
+							clevel,
+							llevel,
+							aclevel-clevel);
+			}
+
+			index++;
+		}
+	}
+
+	/* Adjust total mem and core request. */
+	*reqcore -= reqcoreleft;
```

Comment:
Yes, it is possible to be negative, for example, 1 GRM container is required, but in order to level up the minimum water level, 3 segments have to have 3 more GRM containers added into the request, which should cause reqcoreleft -2.

I will check infinite loop problem and fix it. Thank you!